fms/models/__init__.py

```diff
@@ -236,6 +237,47 @@ def get_model(
                 See `serialization.list_sources(architecture)`
     group: ProcessGroup The PG to use for any model distribution
     """
+    if architecture is None:
+        if variant is not None:
+            logging.warning(
+                f"architecture was set to None which implies the model configuration is being inferred (variant={variant} will be ignored)"
+            )
+
+        if model_path is None:
+            raise ValueError(
+                "the model_path parameter is required to be set in order to infer the model configuration when architecture=None"
+            )
+
+        logging.info(f"inferring model configuration from checkpoint at {model_path}")
+
+        if source is None:
+            raise ValueError(
+                "the model source parameter is required to be set in order to infer the model configuration when architecture=None"
+            )
+        elif source == "hf":
+            if len(kwargs) > 0:
+                logging.warning(
+                    f"ignoring the following parameters as the configuration is being inferred: {list(kwargs.keys())}"
+                )
+            from fms.models.hf import as_fms_model  # type: ignore
+
+            return as_fms_model(
```

Comment:
Instead of recursively calling `get_model` through `as_fms_model` here, can we move the architecture/config guessing code out of `as_fms_model` and have both `get_model` and `as_fms_model` call into that code? Maybe this solution is shorter, but I feel having the guessing code made into a function will make for an easier to understand piece of code.

Reply:
much of the error handling code along this path could also probably be moved to a helper function